// ui-automation/Filter.js
/**
 * @class     Filter
 * @brief     Filter management API
 * @details   Automates the filtering features
 * @authors   Mike Wong mikewong@sfsu.edu
 * @ingroup   MVC
 */

let Filter = (superclass) => class extends superclass {

	// ============================================================
	constructor( page, browser, options ) {
	// ============================================================
		super( page, browser, options );
		this.filter = {
			article: {
				is:      async ( option ) => { await this.filterIs();  return this.filterArticle( option ); },
				not:     async ( option ) => { await this.filterNot(); return this.filterArticle( option ); },
				options: async ()         => { return this.filterArticleOptions(); }
			},
			dgr: {
				is:      async ( option ) => { await this.filterIs();  return this.filterDGR( option ); },
				not:     async ( option ) => { await this.filterNot(); return this.filterDGR( option ); },
				options: async ()         => { return this.filterDGROptions(); }
			},
			excerpt: {
				is:      async ( option ) => { await this.filterIs();  return this.filterExcerpt( option ); },
				not:     async ( option ) => { await this.filterNot(); return this.filterExcerpt( option ); },
			},
			journal: {
				is:      async ( option ) => { await this.filterIs();  return this.filterJournal( option ); },
				not:     async ( option ) => { await this.filterNot(); return this.filterJournal( option ); },
				options: async ()         => { return this.filterJournalOptions(); }
			},
			remove: async ( filter ) => { await this.filterRemove( filter ); }
		};
		this.filters = {
			applied: () => { return this.filtersApplied(); }
		};
	}

	// ============================================================
	async filterIs()  { await this.click( 'input[type="radio"][name="isnot"][value="is"]' );  }
	async filterNot() { await this.click( 'input[type="radio"][name="isnot"][value="not"]' ); }
	// ============================================================

	// ============================================================
	async filtersApplied() {
	// ============================================================
		return await this.page.$$eval( '.filters .filter-item', ( filters ) => { 
			return filters.map(( filter ) => {
				let id  = $( filter ).attr( 'id' );
				let att = $( filter ).children( '.attribute' ) .text();
				let is  = $( filter ).children( '.is' )        .text();
				let val = $( filter ).children( '.value' )     .text();
				return { id: id, attribute: att, is: is, value: val };
			})
		});
	}

	// ============================================================
	async filterRemove( filter ) {
	// ============================================================
		return this.click( `#${filter.id} .remove` );
	}

	// ============================================================
	async filterArticle( option ) {
	// ============================================================
		await this.click( '.filter-select' );
		await this.page.select( '.filter-select', 'Article' );
		await this.page.waitForSelector( `.filter-dropdown option[value="${option}"]` );
		await this.click( '.filter-dropdown' );
		await this.page.select( '.filter-dropdown', option );
		return this.click( '.input-group-btn .btn' );
	}

	// ============================================================
	async filterArticleOptions() {
	// ============================================================
		await this.page.select( '.filter-select', 'Article' );
		return await this.page.evaluate(() => {
			return $( '.filter-dropdown option' ).map(( i, el ) => $( el ).val());
		});
	}

	// ============================================================
	async filterDGR( symbol ) {
	// ============================================================
	// DGR filter drop-down options have JSON data (with DGR IDs) for values
	// but symbol names for the user (since users understand symbol names).
	// Therefore we poll the options for their values and map the symbol names
	// with these JSON data values to enable Puppeteer's select-by-value.
	// ------------------------------------------------------------
		await this.click( '.filter-select' );
		await this.page.select( '.filter-select', 'DGR' );
		await this.page.waitForSelector( '.filter-dropdown' );

		let options = await this.page.evaluate(() => { return [ ... $( '.filter-dropdown option' )].reduce(( options, el ) => { let value = $( el ).val(); let dgr = JSON.parse( value ); options[ dgr.symbol ] = value; return options; }, {}); });
		let value   = options[ symbol ];

		await this.click( '.filter-dropdown' );
		await this.page.select( '.filter-dropdown', value );
		return this.click( '.input-group-btn .btn' );
	}

	// ============================================================
	async filterDGROptions() {
	// ============================================================
		await this.page.select( '.filter-select', 'DGR' );
		return await this.page.evaluate(() => {
			return $( '.filter-dropdown option' ).map(( i, el ) => $( el ).val());
		});
	}

	// ============================================================
	async filterExcerpt( option ) {
	// ============================================================
		await this.click( '.filter-select' );
		await this.page.select( '.filter-select', 'Excerpt' );
		await this.click( '.filter-text' );
		await this.type( option );
		return this.click( '.input-group-btn .btn' );
	}

	// ============================================================
	async filterJournal( option ) {
	// ============================================================
		await this.click( '.filter-select' );
		await this.page.select( '.filter-select', 'Journal' );
		await this.page.waitForSelector( `.filter-dropdown option[value="${option}"]` );
		await this.click( '.filter-dropdown' );
		await this.page.select( '.filter-dropdown', option );
		return this.click( '.input-group-btn .btn' );
	}

	// ============================================================
	async filterJournalOptions() {
	// ============================================================
		await this.page.select( '.filter-select', 'Journal' );
		return await this.page.evaluate(() => {
			return $( '.filter-dropdown option' ).map(( i, el ) => $( el ).val());
		});
	}

}

module.exports = Filter;
