// ui-automation/view/Network.js
class Network {
	constructor() {
		this.nodes = new Set();
		this.edges = new Set();
	}

	// ============================================================
	add_edge( a, b ) {
	// ============================================================
		this.nodes.add( a );
		this.nodes.add( b );

		let edge = new Set([ a, b ]);
		this.edges.add( edge );
	}

	// ============================================================
	hop( a, b, n ) {
	// ============================================================
		let path    = [];
		let start   = [ a ];
		let paths   = this.traverse( path, start, n ); // All paths in graph starting with A of length n or less
		console.log( 'PATHS', paths );
		let nHops   = paths
			.filter( path => path.includes( b ))                                            // Constrain to paths that include destination B
			.map( path => { path.splice( path.indexOf( b ) + 1 ); return path; })           // 
			.map( path => JSON.stringify( path ))
			.filter(( value, index, array ) => { return array.indexOf( value ) === index }) // Constrain to unique values
			.map( path => JSON.parse( path ));

		return nHops;
	}

	oneHop( a, b )   { return this.hop( a, b, 1 ); }
	twoHop( a, b )   { return this.hop( a, b, 2 ); }
	threeHop( a, b ) { return this.hop( a, b, 3 ); }

	// ============================================================
	traverse( path, start, hops ) {
	// ============================================================
		if( path.length >= hops + 1 ) { return [ path ]; }

		let paths = [];
		for( let a of start ) {
			let stop = [ ... this.edges ]
				.filter( edge => edge.has( a ))                          // Constrain to edges that have node A
				.map(    edge => [ ... edge ].find( node => node != a )) // Transform each edge to the node that is not node A
				.filter( node => ! path.includes( node ));               // Constrain to nodes not already on the path

			paths = paths.concat( this.traverse( path.concat( a ), stop, hops ));
		}

		return paths;
	}
}

module.exports = Network;
